### PR TITLE
parse_str() might return arrays by-ref

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/8356

### CODE REVIEW

File: tests/PHPStan/Rules/Arrays/NonexistentOffsetInArrayDimFetchRuleTest.php [hunk 1]
```diff
@@ -700,4 +700,15 @@ public function testBug6243(): void
 		$this->analyse([__DIR__ . '/data/bug-6243.php'], []);
 	}
 
+	public function testBug8356(): void
+	{
+		$this->bleedingEdge = true;
+		$this->analyse([__DIR__ . '/data/bug-8356.php'], [
+			[
+				"Offset 'x' might not exist on array|string.",
```
Comment (staabm) [inline tests/PHPStan/Rules/Arrays/NonexistentOffsetInArrayDimFetchRuleTest.php:708]: before this PR, the rule reported 
`Offset 'x' does not exist on array|string.`
.. now we get  ..
`Offset 'x' might not exist on array|string.`

when running this test without `$this->bleedingEdge = true;` the signature changes in the stub do not have any effect and we still get `Offset 'x' does not exist on array|string.`

**Review by staabm — COMMENTED** [review]
note: the issue-bot did not realize a change in the origin issue snippet, even though the testsuite shows that the PR will change it

File: stubs/core.stub [hunk 2]
```diff
@@ -70,10 +70,16 @@ function bin2hex(string $string): string {}
 
 /**
  * @param array<mixed> $result
- * @param-out array<int|string, string|list<string>> $result
+ * @param-out array<int|string, string|list<string>|array<mixed>> $result
  */
 function parse_str(string $string, array &$result): void {}
```
Comment (staabm) [inline stubs/core.stub:75]: moved `parse_str` and `mb_parse_str` next to each other, and equalized the type signatures

File: stubs/core.stub [hunk 3]
```diff
@@ -70,10 +70,16 @@ function bin2hex(string $string): string {}
 
 /**
  * @param array<mixed> $result
- * @param-out array<int|string, string|list<string>> $result
+ * @param-out array<int|string, string|list<string>|array<mixed>> $result
```
Comment (staabm) [inline stubs/core.stub]: `int` keys are only possible in php7 when using `parse_str($str);` without a second arg.
[per php manual this is highly discouraged](https://www.php.net/manual/en/function.parse-str.php) and was removed with php8:

> Warning
> Using this function without the result parameter is highly DISCOURAGED and DEPRECATED as of PHP 7.2. As of PHP 8.0.0, the result parameter is mandatory.

`mb_parse_str` in contrast does not at all have this feature.

we could drop the `int` part from the array key to get a more precise type for the sugged use-cases, but would get errors for the discouraged use-case in return. wdyt?

File: tests/PHPStan/Rules/Arrays/NonexistentOffsetInArrayDimFetchRuleTest.php (hunk 1 above)
Comment (ondrejmirtes) [reply tests/PHPStan/Rules/Arrays/NonexistentOffsetInArrayDimFetchRuleTest.php:708]: The bot only goes through triaged issues labeled as bug or feature-request.

File: stubs/core.stub (hunk 3 above)
Comment (ondrejmirtes) [reply stubs/core.stub]: `string|list<string>|array<mixed>` - this doesn't make sense to me, that type is going to be normalized: https://phpstan.org/r/7006b26a-88ec-4766-ab84-bc4c8b93160a

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**